docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,518 @@
+*only* the object, not cluster-wide or other constraints, which must be validate
+elsewhere.
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validation=minmax{<numeric value>, <numeric value>}`
+* `// +validation=regexp/<regexp>/`
+* `// +validation=<exact value>`
+
+If no validation directive is in the comments, a stub with a
+`panic("write me!")` marker will be generated. To prevent this, a
+`// +validation=none` directive may be added to the comment.
+
+(We can hash out the exact form of the tags later; it should be easy for human
+readers to figure out what will happen.)
+
+### Package: `apis/group/v1beta1`
+
+This package layout is the same as `apis/group/v1`, but it is an older version
+of the API group, which may lack features.
+
+### Package: `apis/group/v2alpha1`
+
+This package layout is the same as `apis/group/v1`, but it is an newer version
+of the API group, which may have extra features. It has `resource3`, for
+example.
+
+It is a rule that a client should never have to pay attention to multiple
+versions of a group at the same time, so v2alpha1 copies all types from v1, even
+if there's no changes.
+
+## Type support package layout
+
+The above is how to arrange the types for a particular API. But we also must
+consider, what Kubernetes packages do you need to import to use the API types?
+
+```
+k8s.io/api_machinery/
+└── typetools
+    ├── convert
+    ├── deep
+    ├── default
+    └── validate
+```
+
+These packages will live in the kubernetes repo (or another repo that the
+kubernetes team will own and maintain). The interface for all the packages is
+similar: there'll be a `Register(...)` function for the install packages to use,
+and some function to perform the action. So, e.g., `convert.Convert(&in, &out)`,
+`bar := deep.Copy(&foo)`, `default.Apply(&object)`,
+`errs := validate.Check(&object)`.
+
+More detailed descriptions belong in another doc when we hash out the
+particulars of the code generation; the important thing here is just to know
+that we'll supply packages with:
+* Minimal import tree
+* Some sort of registration method. Registration marks a thing as available and
+  it can be enumerated (for producing lists of what could be turned on, for
+  example).
+* Some sort of enabling method, for actually turning a thing on. Registration
+  and enabling are conflated in our codebase at the moment, and this makes it
+  difficult to, for example, construct the default value for --runtime-config.
+* Some sort of execution method, for actually using a thing.
+
+## Reusable apiserver infrastructure packages
+
+In the future, it'd be nice if we could generate REST install code based on
+comment annotations in the go types. See #16560 for a problem statement.
```

Comment:
Nit: could you add the link to https://github.com/kubernetes/kubernetes/issues/16560 here? Github doesn't automatically generate the links in .md.